File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    
    static String check="";
    static String[] anan; 
    
    public String make(List<String> sum){
        String now="";
        for (String s:sum){
            now+=s;
        }
        return now;
    }
    public void make_array(List<String> sum){
        
        int now=0;
        for(String s:sum){
            anan[now]=s;
            now+=1;
        }
    }
    
    
    public void dfs(String[][] tickets,int visit[],List<String> sum,String before){
        int temp=0;
        for(int i=0;i<visit.length;i++){
            if(visit[i]==1)
                temp+=1;
        }
        if(temp==visit.length)
        {
            if(check.length()==0)
            {
                check=make(sum);
                make_array(sum);
            }
            else{
                String s1=make(sum);
                if(s1.compareTo(check)<0){
                    check=s1;
                    make_array(sum);
                }
            }
            return;
        }
        
        for(int i=0;i<tickets.length;i++){
            if(visit[i]==0){
                String s1=tickets[i][0];
                if(before.equals(s1)){
                    visit[i]=1;
                    sum.add(tickets[i][1]);
                    dfs(tickets,visit,sum,tickets[i][1]);
                    visit[i]=0;
                    sum.remove(sum.size() - 1);
                }  
            }
        }    
        
    }
    


    
    public String[] solution(String[][] tickets) {
        String[] answer = {};
        anan=new String[tickets.length+1];
        for(int i=0;i<tickets.length;i++){
            int visit[]=new int[tickets.length];
            String s1=tickets[i][0];
            List<String>an=new LinkedList<>();
            if(s1.equals("ICN")){
                visit[i]=1;
                an.add("ICN");
                an.add(tickets[i][1]);
                dfs(tickets,visit,an,tickets[i][1]);
            }  
            
        }
        
        
        
        return anan;
    }
    
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py




def dfs(visit,graph,now,routes,count):
    if count==len(graph):    
        return routes
    for i in range(len(graph)):
        a1,a2=graph[i]
        if a1==now and visit[i]==0:
            visit[i]=1
            routes.append(a2)
            temp=dfs(visit,graph,a2,routes,count+1)
            if len(temp)!=0:
                return temp
            visit[i]=0
            routes.pop()
    
    return []

def solution(tickets):
    answer = []
    tickets=sorted(tickets,key=lambda x:(x[0],x[1]))
    
    for i in range(len(tickets)):
        a1,a2=tickets[i]
        if a1!='ICN':
            continue
        visit=[0]*len(tickets)
        visit[i]=1
        now=dfs(visit,tickets,a2,[a1,a2],1)
        if len(now)==0:
            continue
        return now
    
